### PR TITLE
feat: add support for image thumbnail

### PR DESCRIPTION
**Pull request by athurg**
While publishing a large number of HD images, the memoe will take a lot of time to load.

So I just generate a smaller thumbnail while uploading image resources, and only load the thumbnail in the memo. If click on the image, it will load the original HD image lazy.

### CODE REVIEW

**Comment by Zeng1998** [comment]
I think dynamic compression when loading the image is even more expensive than loading the full image directly.

**Review by Zeng1998 — COMMENTED** [review]
Do we need to delete the thumbnail in the delete api?

https://github.com/usememos/memos/blob/main/server/resource.go#L273

**Comment by athurg** [comment]
> Do we need to delete the thumbnail in the delete api?
> 
> https://github.com/usememos/memos/blob/main/server/resource.go#L273

You're right, I have fixed it.

File: server/resource.go [hunk 1]
```diff
@@ -127,24 +129,40 @@ func (s *Server) registerResourceRoutes(g *echo.Group) {
...
+
+			fileBytes, err := io.ReadAll(sourceFile)
 			if err != nil {
-				return echo.NewHTTPError(http.StatusInternalServerError, "Failed to create file").SetInternal(err)
+				return echo.NewHTTPError(http.StatusInternalServerError, "Failed to read resource file").SetInternal(err)
 			}
-			defer dst.Close()
-			_, err = io.Copy(dst, sourceFile)
+
+			err = os.WriteFile(filePath, fileBytes, 0666)
```
Comment (Zeng1998) [inline server/resource.go:145]: We don't need to modify the original filename.  It should be handled by the path template.

**Review by Zeng1998 — CHANGES_REQUESTED** [review]

**Review by Zeng1998 — COMMENTED** [review]
I still recommend that it is best not to modify the original filename, this will break the path template settings.

**Comment by athurg** [comment]
> I still recommend that it is best not to modify the original filename, this will break the path template settings.

I have changed the thumbnail's naming. Now the original filename was untouched, and the thumbnail file store in a separate cache dir with resource's PublicID as it's filename.

**Review by Zeng1998 — APPROVED** [review]
LGTM

**Comment by Jonnyan404** [comment]
> > 我还是建议最好不要修改原来的文件名，这样会破坏路径模板的设置。
> 
> 我已经更改了缩略图的命名。现在原始文件名没有改变，缩略图文件存储在一个单独的缓存目录中，文件名是资源的 PublicID。

为啥昨晚的版本更新,我的不显示缩略图?
还是说必须新上传的才能显示?

**Comment by athurg** [comment]
> > > 我还是建议最好不要修改原来的文件名，这样会破坏路径模板的设置。
> > 
> > 
> > 我已经更改了缩略图的命名。现在原始文件名没有改变，缩略图文件存储在一个单独的缓存目录中，文件名是资源的 PublicID。
> 
> 为啥昨晚的版本更新,我的不显示缩略图? 还是说必须新上传的才能显示?

It's not a bug. The Resource Dashboard doesn't support the show the thumbnail before, it only shows an icon.

I have add thumbnail support for the Resource Dashboard, please move to #1666 .

**Pull request merged**